File: examples/basic.py
"""Run after `make bootstrap && make py`: `.venv/bin/python examples/basic.py`."""

from frostwork import Page

html = b"""
<main>
  <h1>Widget</h1>
  <span class="price">$9</span>
  <img src="/a.png"><img src="/b.png">
</main>
"""

page = (
    Page()
    .field("title", "h1::text")
    .field("price", ".price::text")
    .field_all("images", "img::attr(src)")
)
page.check().raise_for_status()
print(page.extract(html).to_json(indent=2))

File: examples/basic.rs
use frostwork::Page;

fn main() {
    let page = Page::new()
        .field("title", "h1::text")
        .field("price", ".price::text")
        .field_all("images", "img::attr(src)");

    // Page caches its compiled plan after this first call and reuses it for later responses.
    let item = page.extract(
        b"<main><h1>Widget</h1><span class=price>$9</span>\
          <img src=/a.png><img src=/b.png></main>",
    );
    println!("{}", item.to_json());
}
